### ISSUE TITLE
Improve OCR by providing regex/info on what to expect?

### ISSUE THREAD

**Issue by cassfalg** [cadon/ARKStatsExtractor#288]
I noticed that OCR recognition is not all that great. In some instances, it recognizes characters that should not appear at that point, like a / instead of a 7.

Can you give hints to the OCR what the characters should be? Like a regex expression? That might help. I could probably create those with some pointers to where to apply them and maybe docs/info on the OCR you use.

**Comment by cadon** [comment]
Hi, the regex for the stats can be found here: https://github.com/cadon/ARKStatsExtractor/blob/master/ARKBreedingStats/ocr/ArkOCR.cs#L723 (`(?:\d+\.\d%?[\/1])?(\d+[\.,']?\d?\d?)(%)?`), if you have an improvement for that, let me know! The ocr for the stats recognizes only letters in this set: `[0-9%/\.\,]` (comma is converted to dot), the mistake `/` instead of `7` is probably the most occuring one.

**Comment by cassfalg** [comment]
That is not quite what I meant I think...

If I understand that code correctly, this is the OCR Call which yields statOCR, the text that OCR recognized?
https://github.com/cadon/ARKStatsExtractor/blob/master/ARKBreedingStats/ocr/ArkOCR.cs#L696
Then later you create the regex and search for matches in what the OCR recognized?

What I thought of doing is to restrict the character set for the OCR routine, something like statOCR = readImage(testbmp, expectedRegex, ...) - so the OCR itself has more information about the text. Not sure if the OCR library can do that of course.

**Comment by cadon** [comment]
Yes, I got you. You mean to tell the OCR something like: do not expect an '/' if you already had one (because the slash appears only once). I think there is no advantage of using this before the point I linked.

Let's think that through with an example: Let the stat to be read be `147.3/270.0`. The OCR doesn't recognizes it correctly and reads `14/.3/2/0.0` (so it reads every '7' as a slash). What would be a good strategy to extract the true value from this? The current regex `(?:\d+\.\d%?[\/1])?(\d+[\.,']?\d?\d?)(%)?` appearantly doesn't get it very well, it gets `14` in the group1. Another issue is, that the dot is not always recognized.
Do you have a suggestion on how to see which slash is the real one, and which other can be ignored or just replaced with a '7'? If you don't know this site yet, I can highly recommend it for testing regexes: https://regex101.com/

I'll think about how to improve this part of the code as well, if you have a neat idea, let me know.

**Comment by cassfalg** [comment]
I didn't really understand that regex yet to be honest. ;) The syntax is a bit different from what I'm used to, I'll have to do some reading on how C# does things.

What's the name of the OCR library? I'd like to read up on it a bit, maybe I'll get some ideas. The thing is, from the OCR results I've seen at a glance, no fancy regex after the OCR is going to fix things. The OCR results are too bad to start with. I've had a number of wrong 6/8/9 recognitions, 5's that were identified as something else... Really strange things.

I didn't use OCR before the UI change - was it better before?

**Comment by cadon** [comment]
I tweaked the ocr again, this time it's doing pretty well. The library was first written by nakram, then adjusted when the UI changed by me. It's in the file https://github.com/cadon/ARKStatsExtractor/blob/master/ARKBreedingStats/ocr/ArkOCR.cs

**Comment by SkyDotNET** [comment]
Another cool webbased regextool im using for almost 10 yrs now is http://txt2re.com
It also generates the code for implementation in almost any popular prpgramming language.
What ilike withthat tool is that you can click on what you want to match in the given source instead of having to build it by yourself (can be problematic in same cases ofc. but most times it was a very handy tool for me).

Btw. the thing with the 7 can be fixed with the custom character selection in the ocr settings.
What i would recommend is to adjust the gamma of the screenshot automatically based on the color of some fixed pixel on the inventory screen, but idk if thats even possible or easy to implement (never did image manipulation with c#)

**Comment by cadon** [comment]
Closing this issue as being similar to #269.
